### PR TITLE
Make benchmark artifact generation a bit more flexible.

### PR DESCRIPTION
**Pull request by ScottTodd**
Here's my use case:

* I'm working on Windows, where we don't support building the TensorFlow integration tools (notably `iree-import-tflite`)
* I want to use the .mlir files that we feed into our benchmarking for local development (more specifically I want _some_ set of real programs, but the benchmark suite offers a set that we continuously build/test and have metrics for already)

---

Now, a sensible solution would be to extend
https://github.com/google/iree/blob/1ddd9170bd6cc2701e2d669754a8fff248739233/build_tools/buildkite/cmake/android/arm64-v8a/benchmark2.yml#L9-L14
to also zip up the imported .mlir files so anyone could download them from Buildkite.

---

I, however, had the more indirect idea to write a Colab notebook ([work in progress here](https://colab.research.google.com/gist/ScottTodd/4b27c6b44f90d239d4676dacddb87895/iree-benchmarks-import-demo.ipynb)) that installs the latest IREE release along with the project source and then builds `iree-benchmark-suites` using the installed tools. That _almost_ worked, except for the few issues that this PR patches over:

* `$<TARGET_FILE:iree::tools::iree-compile>` only works when building with `IREE_BUILD_COMPILER`, but it's also useful to point at already built tools using `IREE_HOST_BINARY_ROOT`. The `iree_get_executable_path` function handles switching between those modes easily
* Compiling imported files for the full list of benchmark configurations can be slow. If a developer just wants to import .mlir files then skipping those following steps would be useful. I added a new `iree-benchmark-import-models` target that does that.

(I also considered the _even more indirect_ idea to parse the artifact paths from [benchmarks/TFLite/CMakeLists.txt](https://github.com/google/iree/blob/main/benchmarks/TFLite/CMakeLists.txt) and run `iree-import-tflite` manually, but the `iree-benchmark-suites` target already does that natively :P)

---

I'm open to any workflow that solves my original use case, but these changes seem generally good to have anyways.

### CODE REVIEW

**Comment by pzread** [comment]
LGTM with my limited knowledge : )

File: build_tools/cmake/iree_benchmark_suite.cmake [hunk 1]
```diff
@@ -163,6 +164,13 @@ function(iree_benchmark_suite)
       set(_MODULE_SOURCE_TARGET "${_TFLITE_IMPORT_TARGET}")
     endif()
 
+    # Allow an early exit for developers that just want to import module
+    # sources rather than fully compile to IREE modules for each configuration.
+    if(IREE_BUILD_BENCHMARKS_IMPORT_ONLY)
+      add_dependencies(iree-benchmark-suites "${_MODULE_SOURCE_TARGET}")
```
[the remark this answers is not in the document]
Comment (ScottTodd) [reply build_tools/cmake/iree_benchmark_suite.cmake:170]: Yeah! I'm not sure what plans there are for this code so didn't want to refactor too much, but splitting into multiple functions and targets would accomplish the same thing without needing to add a new option. `iree-benchmark-suites` would depend on `iree-benchmark-import-models` and developers / scripts could choose to build either.

Also looking at this again, I'm not sure how much we gain from having this benchmark artifact fetching/importing/compiling/etc. all in CMake vs having a set of python/bash scripts. A 250+ line CMake function can be pretty tricky to maintain.

[the remark this answers is not in the document]
Comment (antiagainst) [reply build_tools/cmake/iree_benchmark_suite.cmake:170]: Actually we [already have](https://github.com/google/iree/blob/main/build_tools/cmake/iree_benchmark_suite.cmake#L156) a dedicated target for importing each TFLite FlatBuffer. It's better to me to define another meta target like `iree-benchmark-import-models` for it. It should be straightforward, just define it [here](https://github.com/google/iree/blob/main/CMakeLists.txt#L552) and make add `add_dependencies(iree-benchmark-import-models "${_TFLITE_IMPORT_TARGET}")` [here](https://github.com/google/iree/blob/main/build_tools/cmake/iree_benchmark_suite.cmake#L161). No need to let `iree-benchmark-suites` depends on `iree-benchmark-import-models` as each model tracks its own dependent chain (in order to allow downloading/importing/compiling a specific model).

**Comment by antiagainst** [comment]
> Now, a sensible solution would be to extend
> 
> https://github.com/google/iree/blob/1ddd9170bd6cc2701e2d669754a8fff248739233/build_tools/buildkite/cmake/android/arm64-v8a/benchmark2.yml#L9-L14
> 
> 
> to also zip up the imported .mlir files so anyone could download them from Buildkite.

This is also fine to me. I was previously worried about the storage usage but I guess Buildkite will compute SHA and dedup on their side. The only cost is the CPU time to zip. That happens in powerful cloud VMs though; shouldn't be too long.

**Review by antiagainst — CHANGES_REQUESTED** [review]
Instead of introducing a CMake option, let's define a CMake target for it.

**Comment by antiagainst** [comment]
> > Now, a sensible solution would be to extend
> > https://github.com/google/iree/blob/1ddd9170bd6cc2701e2d669754a8fff248739233/build_tools/buildkite/cmake/android/arm64-v8a/benchmark2.yml#L9-L14
> > 
> > to also zip up the imported .mlir files so anyone could download them from Buildkite.
> 
> This is also fine to me. I was previously worried about the storage usage but I guess Buildkite will compute SHA and dedup on their side. The only cost is the CPU time to zip. That happens in powerful cloud VMs though; shouldn't be too long.

It does make our development easier sometimes. Trying out with https://github.com/google/iree/pull/8858. :)

File: build_tools/cmake/iree_benchmark_suite.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (ScottTodd) [reply build_tools/cmake/iree_benchmark_suite.cmake]: Ah, good point. I updated this as you suggested. PTAL.

Also, here's the Colab notebook I'm hacking on: https://colab.research.google.com/gist/ScottTodd/4b27c6b44f90d239d4676dacddb87895/iree-benchmarks-import-demo.ipynb. It takes a few minutes to run when just importing the files (most of the time is actually spent updating submodules so the CMake configure succeeds). With a little bit more work, the notebook could also automatically zip and download the imported files (for now you have to click through the file browser and download each individually)

File: CMakeLists.txt [hunk 2]
```diff
@@ -549,8 +549,9 @@ endif()
 #-------------------------------------------------------------------------------
 
 if(${IREE_BUILD_BENCHMARKS})
-  # Add a top-level custom target to drive generating benchmark suites.
+  # Add top-level custom targets to drive generating benchmark suites.
```
Comment (antiagainst) [inline CMakeLists.txt:552]: Add some more explanation about what these targets are for?

**Review by antiagainst — APPROVED** [review]

**Pull request merged**